### ISSUE TITLE
Mixing mainnet and testnet addresses in outputs

### ISSUE THREAD

**Issue by bjarnemagnussen** [ofek/bit#34]
Bitcoin mainnet and testnet addresses encode the same information/scripts, but have different prefixes to differentiate between the two networks.

Currently in Bit it is possible to e.g. spend from a mainnet address/PrivateKey but include testnet addresses in the outputs (and vice versa, and mixed cases). Bit then simply converts those testnet addresses to its mainnet equivalent in the outputs of the spending transaction. 

This may lead to confusion, and I would like to discuss if maybe there should be some sanity checks in place to let the user know when a mix of mainnet and testnet addresses and private keys are used?

**Comment by bjarnemagnussen** [comment]
Where in the code should we test for that?

One suggestion is to add a check within the function ``sanitize_tx_data`` to make sure that when ``PrivateKey`` calls it, then all output addresses must be for mainnet; and when ``PrivateKeyTestnet`` calls it, then all output addresses must be for testnet: 

```
diff --git a/bit/transaction.py b/bit/transaction.py
index b16cf79..81209c5 100644
--- a/bit/transaction.py
+++ b/bit/transaction.py
@@ -4,7 +4,7 @@ from itertools import islice
 
 from bit.crypto import double_sha256, sha256
 from bit.exceptions import InsufficientFunds
-from bit.format import address_to_public_key_hash
+from bit.format import address_to_public_key_hash, get_version
 from bit.network.rates import currency_to_satoshi_cached
 from bit.utils import (
     bytes_to_hex, chunk_data, hex_to_bytes, int_to_unknown_bytes
@@ -78,7 +78,7 @@ def estimate_tx_fee(n_in, n_out, satoshis, compressed):
     return estimated_fee
 
 
-def sanitize_tx_data(unspents, outputs, fee, leftover, combine=True, message=None, compressed=True):
+def sanitize_tx_data(unspents, outputs, fee, leftover, combine=True, message=None, compressed=True, version='main'):
     """
     sanitize_tx_data()
 
@@ -89,6 +89,12 @@ def sanitize_tx_data(unspents, outputs, fee, leftover, combine=True, message=Non
 
     for i, output in enumerate(outputs):
         dest, amount, currency = output
+        # Sanity check: If spending from main-/testnet, then all output addresses must also be for main-/testnet.
+        if amount:  # ``dest`` could be a text to be stored in the blockchain; but only allowed if ``amount`` is exactly zero.
+            vs = get_version(dest)
+            if vs != version:
+                raise ValueError('Cannot send to ' + vs + 'net address when spending from a ' + version + 'net address.')
+
         outputs[i] = (dest, currency_to_satoshi_cached(amount, currency))
 
     if not unspents:
```

The function must now be called with a ``version`` parameter:

```
diff --git a/bit/wallet.py b/bit/wallet.py
index c6167d9..fb66f64 100644
--- a/bit/wallet.py
+++ b/bit/wallet.py
@@ -234,7 +234,8 @@ class PrivateKey(BaseKey):
             leftover or self.address,
             combine=combine,
             message=message,
-            compressed=self.is_compressed()
+            compressed=self.is_compressed(),
+            version='main'
         )
 
         return create_p2pkh_transaction(self, unspents, outputs)
@@ -329,7 +330,8 @@ class PrivateKey(BaseKey):
             leftover or address,
             combine=combine,
             message=message,
-            compressed=compressed
+            compressed=compressed,
+            version='main'
         )
 
         data = {
@@ -518,7 +520,8 @@ class PrivateKeyTestnet(BaseKey):
             leftover or self.address,
             combine=combine,
             message=message,
-            compressed=self.is_compressed()
+            compressed=self.is_compressed(),
+            version='test'
         )
 
         return create_p2pkh_transaction(self, unspents, outputs)
@@ -613,7 +616,8 @@ class PrivateKeyTestnet(BaseKey):
             leftover or address,
             combine=combine,
             message=message,
-            compressed=compressed
+            compressed=compressed,
+            version='test'
         )
 
         data = {
```

**These changes do break numerous tests in ``test/test_transaction.py``.**

**Comment by teran-mckinney** [comment]
This is definitely a good thing to have. Personally, I think sanitize_tx_data and just having it compare the outputs, if a different address type is on both, raise a ValueError. I think that will break fewer tests, can still be tested easily, and isn't overly invasive. I don't see why a user would ever want to attempt mixing mainnet and testnet addresses.

I would keep the assumption of which version to the addresses themselves. Not sure off the top of my head if outputs are typed and the type can return the chain. That could be a very explicit way for a user to implement it, if he or she wanted. `outputs = ('mainnet_address', 10000)`, `assert outputs.target_chain == 'mainnet'.`, etc. Also forgot if outputs are bundled or each have individual entries, so possibly that area could be another spot to implement it in.

Thanks for spotting this.

**Comment by bjarnemagnussen** [comment]
Fixed with commit 95098aa3c08b937f12e1c3238c9de6817d570401
